### PR TITLE
Prevent lesson editor from adding the same level to a script twice

### PR DESCRIPTION
**Pull request by davidsbailey**
Finishes [PLAT-612]. 

It's hard to anticipate whether a level is already in the script or not before updating the lesson, so we use the existing database transaction to check that there are no duplicates after the lesson has been updated, but before the transaction is committed.

## Testing story

New unit test verifies that the error case is hit on the server. I also manually verified that the error is visible to the user when adding a duplicate level via the UI:
Screen Shot 2021-02-14 at 6 33 07 AM

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

[PLAT-612]: https://codedotorg.atlassian.net/browse/PLAT-612

### CODE REVIEW

File: dashboard/test/controllers/lessons_controller_test.rb [hunk 1]
```diff
@@ -779,10 +751,36 @@ class LessonsControllerTest < ActionController::TestCase
     script_level = section.script_levels.first
     assert script_level.assessment
     refute script_level.bonus
-    assert_equal 'progression name', script_level.progression
```
Comment (dmcavoy) [inline dashboard/test/controllers/lessons_controller_test.rb]: Why are you removing this?

**Review by dmcavoy — APPROVED** [review]
Looks good! I like how you improved the test while you were there. One question since I think we still use progression.

File: dashboard/app/models/script.rb [hunk 2]
```diff
@@ -114,6 +114,17 @@ def hide_pilot_scripts
 
   validate :set_is_migrated_only_for_migrated_scripts
 
+  def prevent_duplicate_levels
+    reload
+
+    unless levels.count == levels.uniq.count
+      h = {}
+      levels.map(&:key).each {|key| h[key] = (h[key] || 0) + 1}
```
Comment (bethanyaconnor) [inline dashboard/app/models/script.rb]: I think you could use group_by to simplify this logic slightly, like `duplicate_keys = levels.map(&:key).group_by {|key| key}.select {|_key, values| values.count > 1}.keys`

But I don't think it makes much of a difference so up to you.

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/test/controllers/lessons_controller_test.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/test/controllers/lessons_controller_test.rb]: good catch... I've added this back.

File: dashboard/app/models/script.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/script.rb]: 💡 ! I have looked up how to do "count by" in ruby many times and never thought to use `group_by`. Thanks for this!

**Review by davidsbailey — COMMENTED** [review]
Thank you both for the careful review!

**Pull request merged**